fc/language/expressions/array.py

```diff
@@ -562,57 +551,61 @@ def Interpret(self, env):
+                                "dimensions, so it cannot be folded along dimension", dim.value)
         if not hasattr(shrink, 'value'):
-                raise ProtocolError("The shrink input should be a simple value, not a", type(shrink))
+            raise ProtocolError("The shrink input should be a simple value, not a", type(shrink))
         if not hasattr(pad, 'value'):
-                raise ProtocolError("The pad input should be a simple value, not a", type(pad))
+            raise ProtocolError("The pad input should be a simple value, not a", type(pad))
         if shrink.value != 0 and pad.value != 0:
             raise ProtocolError("You cannot both pad and shrink!")
         if not hasattr(pad, 'value'):
-                raise ProtocolError("The pad_value input should be a simple value, not a", type(pad_value))
+            raise ProtocolError("The pad_value input should be a simple value, not a", type(pad_value))
 
         dim_val = int(dim.value)
-        shape = list(operand.array.shape)
+        shape = list(map(int, operand.array.shape))
         num_entries = indices.array.shape[0]
         shape[dim_val] = 1
-        extents = np.zeros(tuple(shape), dtype=float)
+        extents = np.zeros(tuple(shape), dtype=np.uint32)
+        indices.array = indices.array.astype(int)
         for index in indices.array:
             extents_index = list(index)
             extents_index[dim_val] = 0
             extents[tuple(extents_index)] += 1
         max_extent = np.amax(extents)
         min_extent = np.amin(extents)
         if min_extent == 0 and pad.value == 0 or (min_extent != max_extent and shrink.value == 0 and pad.value == 0):
-            raise ProtocolError("Cannot index if the result is irregular (extent ranges from", min_extent, "to", max_extent, ").")
+            raise ProtocolError("Cannot index if the result is irregular (extent ranges from",
+                                min_extent, "to", max_extent, ").")
         if pad.value != 0:
             extent = max_extent
         else:
-            extent = min_extent 
+            extent = min_extent
         shape[dim_val] = extent
-            
+
         result = np.empty(shape)
-        
+
         if pad != 0:
             result.fill(pad_value)
         if shrink.value + pad.value < 0:
-            begin = num_entries - 1   
+            begin = num_entries - 1
             end = -1
             move = -1
         else:
             begin = 0
             end = num_entries
             move = 1
-        
-        # The next_index array keeps track of how far along dimension dim_val we should put the next kept entry at each location
+
+        # The next_index array keeps track of how far along dimension dim_val we
+        # should put the next kept entry at each location
         shape[dim_val] = 1
-        next_index = np.zeros(shape)
+        next_index = np.zeros(shape, dtype=np.uint32)
```

Comment:
Any reason for this particular type?

Reply:
It needs to be an integer type so we can use it to index numpy arrays, and it doesn't need to be negative, so I went for 32-bit unsigned

Reply:
Ok. Nothing wrong with that! Only it makes it look like there's going to be some sort of interaction with low-level code at one point, or maybe packing into a file or such. Would there be any downside to just writing dtype=int ?